proposals/0000-scoped-type-variable-puns.md

```diff
@@ -0,0 +1,137 @@
+---
+author: Georgi Lyubenov
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/507).
+
+# Allow puns to bind type variables
+
+The current language extensions `NamedFieldPuns` and `ScopedTypeVariables` have a (surprisingly) missing interaction:
+
+```haskell
+data Thing a = MkThing {field :: a}
+
+f :: Thing a -> a
+f MkThing {field :: x} = undefined :: x
+```
+
+This currently fails to parse. I believe that a natural extension of the interaction the two extensions
+would be to instead have the above mean the same as
+
+```haskell
+data Thing a = MkThing {field :: a}
+
+f :: Thing a -> a
+f MkThing {field = field :: x} = undefined :: x
+```
+
+## Motivation
+
+`NamedFieldPuns` and `ScopedTypeVariables` are both firmly cemented as useful and ubiquitous language extensions.
+As we move towards improved type level programming, we should aim to make the UX as seamless as possible.
+I think that allowing the `MkField {field :: x}` syntax would be considered "obvious" and "natural" to anyone who has used both extensions,
+to the point where when I tried it I had no doubt it would work.
+
+Apart from the obvious convenience that this would provide, it is also less error-prone(as is usual with using field names instead of positional matches) compared to binding type variables in patterns, as described [here](https://gitlab.haskell.org/ghc/ghc/-/issues/18830#note_308891).
+
+## Proposed Change Specification
+
+I propose that a pattern match of the form
+```haskell
+Foo {bar :: baz}
+```
+be parsed successfully and desugared(in the renamer) to
+```haskell
+Foo {bar = bar :: baz}
+```
+
+Example grammar change from `ghc`'s `Parser.y`:
+Current:
+```
+fbind
+        : qvar '=' texp
+        | qvar
+        ...
+```
+Proposed addition:
+```
+fbind
+        : qvar '=' texp
+        | qvar '::' ctype
+        | qvar
+        ...
+```
+
+TODO: I'm not sure if the Haskell report contains a specification for the grammar of field puns and scoped type variables.
+If anyone wants to point me to where those are specified, I'd be happy to update this section.
+
+TODO: I'm not sure what corner cases might exist for this, since it is only a syntax sugar change.
+Following this, the only things I can imagine are parser issues.
+
+The specification can, and almost always should, be illustrated with
+*examples* that illustrate corner cases. But it is not sufficient to
+give a couple of examples and regard that as the specification! The
+examples should illustrate and elucidate a clearly-articulated
+specification that covers the general case.
+
+## Examples
+
+```haskell
+data Thing a = MkThing {field :: a}
+
+-- x is able to be referred to in the body here, in the same way in which it would be if we wrote
+-- MkThing {field = field :: x}
+f :: Thing a -> ()
+f MkThing {field :: x} = ()
+```
+
+## Effect and Interactions
+
+TODO: not sure what to write here.
```

Comment:
I agree there isn't much to write here.

Reply:
Presumably this'll interact with `-XExplicitForall` and `-XScopedTypeVariables` in that in

```haskell
    f :: forall a. Thing a -> a
    f MkThing{field :: a} = undefined :: a
```

the `a` appearing in the field annotation is the `a` `forall`'d in `f`'s signature.

(And that variety of scoping is contra one of the Principles for future GHC, [point 4. here](https://github.com/goldfirere/ghc-proposals/blob/type-variables/principles.rst#224contiguous-scoping-principle-csp).)

Reply:
This is however the same interaction as what is currently present if you write `MkThing {field = field :: a}`, so I'm not sure if I should mark it down as an interaction of this proposal.